File: src/test/scala/com/exasol/cloudetl/bucket/SecureBucketTest.scala
package com.exasol.cloudetl.bucket

import com.exasol.ExaMetadata
import com.exasol.cloudetl.storage.StorageProperties

import org.apache.hadoop.conf.Configuration

class SecureBucketTest extends AbstractBucketTest {

  test("validate throws if no connection name is provided") {
    assertNoConnectionName(BaseSecureBucket(properties).validate())
  }

  test("validate throws if access properties are provided") {
    properties = Map(
      "accountNameProperty" -> "user",
      "accountSecretProperty" -> "secret"
    )
    assertForbiddenProperty(BaseSecureBucket(properties).validate())
  }

  private[this] case class BaseSecureBucket(
    val params: Map[String, String],
    val metadata: ExaMetadata = null
  ) extends Bucket
      with SecureBucket {
    override val properties = StorageProperties(params, metadata)

    override val bucketPath = "local_path"
    override def getRequiredProperties(): Seq[String] = Seq.empty[String]
    override def getSecureProperties(): Seq[String] = Seq("accountSecretProperty")
    override def getConfiguration(): Configuration = new Configuration()
    override def validate(): Unit = {
      validateRequiredProperties()
      validateConnectionProperties()
    }
  }
}

File: src/test/java/com/exasol/cloudetl/bucket/SecureBucketTest.java
package com.exasol.cloudetl.bucket;

import java.util.List;
import java.util.Map;

import org.apache.hadoop.conf.Configuration;
import org.junit.jupiter.api.Test;

import com.exasol.ExaMetadata;
import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.storage.StorageProperties;

class SecureBucketTest extends AbstractBucketTest {
    @Test
    void validateThrowsIfNoConnectionNameIsProvided() {
        assertNoConnectionName(() -> new BaseSecureBucket(this.properties, null).validate());
    }

    @Test
    void validateThrowsIfAccessPropertiesAreProvided() {
        this.properties = Map.of("accountNameProperty", "user", "accountSecretProperty", "secret");
        assertForbiddenProperty(() -> new BaseSecureBucket(this.properties, null).validate());
    }

    private static final class BaseSecureBucket extends Bucket implements SecureBucket {
        private final StorageProperties properties;

        private BaseSecureBucket(final Map<String, String> params, final ExaMetadata metadata) {
            this.properties = metadata == null ? new StorageProperties(params) : new StorageProperties(params, metadata);
        }

        @Override
        public StorageProperties properties() {
            return this.properties;
        }

        @Override
        public String bucketPath() {
            return "local_path";
        }

        @Override
        public scala.collection.immutable.Seq<String> getRequiredProperties() {
            return ScalaConverters.seqFromJava(List.of());
        }

        @Override
        public scala.collection.immutable.Seq<String> getSecureProperties() {
            return ScalaConverters.seqFromJava(List.of("accountSecretProperty"));
        }

        @Override
        public Configuration getConfiguration() {
            return new Configuration();
        }

        @Override
        public void validate() {
            validateRequiredProperties();
            validateConnectionProperties();
        }
    }
}
